File: 08. Trees/10. Zigzag Traversal/main.js
class TreeNode {
    constructor(val) {
        this.val = val;
        this.left = this.right = null;
    }
}

// BFS Approach
function zigzagLevelOrderBFS(root) {
    if (!root) return [];
    
    let result = [], queue = [root], leftToRight = true;
    
    while (queue.length) {
        let size = queue.length, level = new Array(size).fill(0);
        
        for (let i = 0; i < size; i++) {
            let node = queue.shift();
            let index = leftToRight ? i : size - 1 - i;
            level[index] = node.val;
            
            if (node.left) queue.push(node.left);
            if (node.right) queue.push(node.right);
        }
        
        result.push(level);
        leftToRight = !leftToRight;
    }
    return result;
}

// DFS Approach
function dfs(node, level, leftToRight, result) {
    if (!node) return;

    if (result.length === level)
        result.push([]);

    if (leftToRight)
        result[level].push(node.val);
    else
        result[level].unshift(node.val);

    dfs(node.left, level + 1, !leftToRight, result);
    dfs(node.right, level + 1, !leftToRight, result);
}

function zigzagLevelOrderDFS(root) {
    let result = [];
    dfs(root, 0, true, result);
    return result;
}

// Main function
function main() {
    let root = new TreeNode(3);
    root.left = new TreeNode(9);
    root.right = new TreeNode(20);
    root.right.left = new TreeNode(15);
    root.right.right = new TreeNode(7);

    console.log("Zigzag Level Order (BFS):", zigzagLevelOrderBFS(root));
    console.log("Zigzag Level Order (DFS):", zigzagLevelOrderDFS(root));
}

main();

File: 08. Trees/10. Zigzag Traversal/main.py
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# BFS Approach
def zigzagLevelOrderBFS(root):
    if not root:
        return []

    result = []
    queue = deque([root])
    left_to_right = True

    while queue:
        level_size = len(queue)
        level = [0] * level_size

        for i in range(level_size):
            node = queue.popleft()
            index = i if left_to_right else level_size - 1 - i
            level[index] = node.val

            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

        result.append(level)
        left_to_right = not left_to_right

    return result

# DFS Approach
def dfs(node, level, left_to_right, result):
    if not node:
        return

    if len(result) == level:
        result.append([])

    if left_to_right:
        result[level].append(node.val)
    else:
        result[level].insert(0, node.val)

    dfs(node.left, level + 1, not left_to_right, result)
    dfs(node.right, level + 1, not left_to_right, result)

def zigzagLevelOrderDFS(root):
    result = []
    dfs(root, 0, True, result)
    return result

# Main function
def main():
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)

    print("Zigzag Level Order (BFS):", zigzagLevelOrderBFS(root))
    print("Zigzag Level Order (DFS):", zigzagLevelOrderDFS(root))

main()
